### PR TITLE
adding more warnings for GCC

### PR DESCRIPTION
**Pull request by kreuzerkrieg**
Check out warnings added for GCC, if agreed I will fix all related compilation problems.
Short synopsis added for each argument for your convenience 

For changelog. Remove if this is non-significant change.

Category (leave one):
- Build/Testing/Packaging Improvement

Short description (up to few sentences):
Added more warnings for GCC build that are not covered by `-Wall -Wextra -Werror`.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Ok, but we don't need to list warnings that are already present in `-Wextra`.

File: dbms/CMakeLists.txt
Comment (alexey-milovidov) [inline dbms/CMakeLists.txt]: I'm not an expert in CMake. I can guess that it means "add this option only if language is C++". Maybe it worth a comment.

File: dbms/CMakeLists.txt
Comment (alexey-milovidov) [inline dbms/CMakeLists.txt]: I'm sure that we already have it :)

File: dbms/CMakeLists.txt
Comment (alexey-milovidov) [inline dbms/CMakeLists.txt]: We already have `-Wall -Wextra -Werror`.

**Comment by kreuzerkrieg** [comment]
I can go over and remove all those included in `extra` and `all`
But since no one remembers what is included in `all` and `extra`, I prefer to state warning switches explicitly. But it is my personal preference, of course. 
Your call.

File: dbms/CMakeLists.txt
Comment (kreuzerkrieg) [reply dbms/CMakeLists.txt]: good idea, will add

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
> But since no one remembers what is included in all and extra, I prefer to state warning switches explicitly. But it is my personal preference, of course.
Your call.

We can find them in gcc docs. (But it's Ok to leave some duplicates...)

**Comment by kreuzerkrieg** [comment]
> > But since no one remembers what is included in all and extra, I prefer to state warning switches explicitly. But it is my personal preference, of course.
> > Your call.
> 
> We can find them in gcc docs. (But it's Ok to leave some duplicates...)

just removed all redundant stuff
if you are ok with these, get your guys ready for float comparison failures, since I have no idea how you would solve it

**Comment by alexey-milovidov** [comment]
> if you are ok with these, get your guys ready for float comparison failures

We are Ok with exact (bitwise) float comparisons in most of cases.
If you don't see something suspicious, we can simply omit this warning.

**Comment by alexey-milovidov** [comment]
More info: https://randomascii.wordpress.com/2012/06/26/doubles-are-not-floats-so-dont-compare-them/

**Comment by kreuzerkrieg** [comment]
> More info: https://randomascii.wordpress.com/2012/06/26/doubles-are-not-floats-so-dont-compare-them/

Yep, read this one the day before yesterday. If you are sure you are ok with `==` and `!=` let it be, I guess you know better what are you doing. On the other hand, why do you have special functions to compare floats?

**Comment by alexey-milovidov** [comment]
> If you are sure you are ok with == and != let it be

Yes. We usually compare floats that was not affected by any floating point math. For example:
```
float x = 0;
... maybe set x to some non-zero value
if (x == 0)
    do something
```
There is not much floating point math in ClickHouse. And if any, we usually compare for less/greater.

> why do you have special functions to compare floats?

We don't have (almost).

There are some edge cases... For example, when we do sorting, we must take care to compare NaNs in a different way (look at ColumnVector.h, struct CompareHelper). When we do GROUP BY, we compare floats for bitwise equality, hash them, etc...

**Comment by kreuzerkrieg** [comment]
> > If you are sure you are ok with == and != let it be
> 
> Yes. We usually compare floats that was not affected by any floating point math. For example:
> 
> ```
> float x = 0;
> ... maybe set x to some non-zero value
> if (x == 0)
>     do something
> ```
> 
> There is not much floating point math in ClickHouse. And if any, we usually compare for less/greater.
> 
> > why do you have special functions to compare floats?
> 
> We don't have (almost).
> 
> There are some edge cases... For example, when we do sorting, we must take care to compare NaNs in a different way (look at ColumnVector.h, struct CompareHelper). When we do GROUP BY, we compare floats for bitwise equality, hash them, etc...

Sounds like the warning could be safely omitted

**Comment by kreuzerkrieg** [comment]
@alexey-milovidov couple of notes
Some files were reformatted, of course, according to your clang format settings, if you strongly oppose the noise it creates let me know, I will revert.
Second, there were some warning suppression, which I strongly oppose, except cases when the noise originated from third party, which is not always the case with CH, please take a look at Source.h. It have a template class which inherits from template argument, sometimes it results in overriding virtual function, sometimes it is not, depends on template instantiation.  Please take a look, I'm not sure what to do with it.
Third, the `registerWhatever()` functions, technically it misses declaration. It is a work in progress, let me know if for some reason you dont want it be declared in header files

More warnings to uncomment and fix compilation errors.

File: dbms/src/Storages/MergeTree/MergeTreeIndices.h
Comment (alexey-milovidov) [inline dbms/src/Storages/MergeTree/MergeTreeIndices.h]: Maybe better to make functions static / in anonymous namespace (internal linkage) instead of exposing in header?

File: libs/libmysqlxx/include/mysqlxx/Exception.h
Comment (alexey-milovidov) [inline libs/libmysqlxx/include/mysqlxx/Exception.h]: :+1:

**Comment by alexey-milovidov** [comment]
That's great. I quickly looked through your modifications and they look very valuable!

> Some files were reformatted, of course, according to your clang format settings, if you strongly oppose the noise it creates let me know, I will revert.

It's Ok unless it changes whole file (namespace indentation or something like that) or perform unpleasant reformatting of array tables. I don't see anything like that in the modifications.

> Second, there were some warning suppression, which I strongly oppose, except cases when the noise originated from third party, which is not always the case with CH, please take a look at Source.h. It have a template class which inherits from template argument, sometimes it results in overriding virtual function, sometimes it is not, depends on template instantiation. Please take a look, I'm not sure what to do with it.

dbms/src/Functions/GatherUtils/Sources.h

Just leave the suppression. In this file we use a base class with virtual methods or another with non-virtual methods for the reason.

> Third, the registerWhatever() functions, technically it misses declaration. It is a work in progress, let me know if for some reason you dont want it be declared in header files

If you found it convenient, it is Ok. And we will follow.

File: dbms/src/IO/IReadableWriteBuffer.h
Comment (kreuzerkrieg) [inline dbms/src/IO/IReadableWriteBuffer.h]: @alexey-milovidov  take a look here
The compiler complains if possible null dereference, do you have any idea why it is not satisfied?

File: dbms/programs/odbc-bridge/CMakeLists.txt
Comment (kreuzerkrieg) [inline dbms/programs/odbc-bridge/CMakeLists.txt]: @alexey-milovidov , here the compiler complains about missing include folder and for reason, by using `set` and multiple (separated by spaces) values creates a list, then it is pushed into `target_include_directory` and values like "blah/foo/bar/baz/ SYSTEM" becomes "blah/foo/bar/baz/SYSTEM" which is not a folder and plainly wrong to do. I've tried to add double quotes to denote a string, but it didnt make the trick. Do you have some CMake guru at your disposal? :)

File: dbms/src/Functions/IFunction.h
Comment (kreuzerkrieg) [inline dbms/src/Functions/IFunction.h]: @alexey-milovidov 
dont like that I dodnt pop the pragma. It means, anything including this header gets the error hidden. Any idea how to solve it?

Comment (alexey-milovidov) [reply dbms/src/Functions/IFunction.h]: We should rename that function.

Comment (kreuzerkrieg) [reply dbms/src/Functions/IFunction.h]: Looks like I will be busy until the end of September. Hopefully will be beck to finish this PR

Comment (alexey-milovidov) [reply dbms/src/Functions/IFunction.h]: Ok, we'll wait.

**Comment by kreuzerkrieg** [comment]
Any ideas whats about Darwin build? Looks like its failed to start

**Comment by alexey-milovidov** [comment]
> Any ideas whats about Darwin build? Looks like its failed to start

It will fix after you merge with master.

**Comment by kreuzerkrieg** [comment]
> Any ideas whats about Darwin build? Looks like its failed to start
> 
> It will fix after you merge with master.

Already fixed

File: dbms/src/Functions/timezone.cpp
Comment (alexey-milovidov) [inline dbms/src/Functions/timezone.cpp]: Not sure this is not an overkill. But let's proceed...

Comment (kreuzerkrieg) [reply dbms/src/Functions/timezone.cpp]: It is MEGA overkill
I'm already afraid of this PR :)

File: dbms/programs/obfuscator/Obfuscator.cpp
Comment (alexey-milovidov) [inline dbms/programs/obfuscator/Obfuscator.cpp]: Function try block is obscure feature but why someone compains?

File: dbms/programs/server/MySQLHandler.cpp
Comment (alexey-milovidov) [inline dbms/programs/server/MySQLHandler.cpp]: What does it mean ("warning" instead of "ignored")?

File: dbms/src/AggregateFunctions/AggregateFunctionGroupBitmapData.h
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionGroupBitmapData.h]: We can include these headers as SYSTEM in CMake. This will imply `-isystem` for compiler option and subsequently, it means "ignore all warnings in this header".

File: dbms/src/Common/OpenSSLHelpers.cpp
Comment (alexey-milovidov) [inline dbms/src/Common/OpenSSLHelpers.cpp]: BTW, we don't want any warning in successfull build (we use `-Werror`) because warnings will steal our attention every time.

**Comment by alexey-milovidov** [comment]
BTW, I found these changes very useful. Especially: forcing functions to have internal linkage, forcing `override`...

**Comment by alexey-milovidov** [comment]
Except "missing-declarations". Probably it'd better be rolled back.

File: dbms/src/DataStreams/materializeBlock.cpp
Comment (alexey-milovidov) [inline dbms/src/DataStreams/materializeBlock.cpp]: Nothing wrong was here. But this is Ok also.

File: dbms/src/Functions/CMakeLists.txt
Comment (alexey-milovidov) [inline dbms/src/Functions/CMakeLists.txt]: Don't understand :(

File: dbms/src/Functions/IFunction.h
Comment (alexey-milovidov) [inline dbms/src/Functions/IFunction.h]: Better to remove.

File: dbms/src/Functions/CMakeLists.txt
Comment (kreuzerkrieg) [reply dbms/src/Functions/CMakeLists.txt]: Temporary stuff with compilation problem of ASAN stuff, we discussed it via Telegram channel

File: dbms/src/Functions/IFunction.h
Comment (kreuzerkrieg) [reply dbms/src/Functions/IFunction.h]: The PR is far from being complete, I do go over my changes in the "final" commit, to remove all these leftovers

File: dbms/src/DataStreams/materializeBlock.cpp
Comment (kreuzerkrieg) [reply dbms/src/DataStreams/materializeBlock.cpp]: Yep, I also dont see any problem as the previous code from my point of view was valid, however, it emitted warning

File: dbms/src/Common/OpenSSLHelpers.cpp
Comment (kreuzerkrieg) [reply dbms/src/Common/OpenSSLHelpers.cpp]: I guess copy/paste "bug" of course it is "ignore"

File: dbms/src/IO/IReadableWriteBuffer.h
Comment (alexey-milovidov) [reply dbms/src/IO/IReadableWriteBuffer.h]: At what exact location possible dereference happens?

**Comment by alexey-milovidov** [comment]
I have read almost all changes and this PR is generally Ok.

File: dbms/programs/obfuscator/Obfuscator.cpp
Comment (kreuzerkrieg) [reply dbms/programs/obfuscator/Obfuscator.cpp]: It was me, I will revert it

**Comment by kreuzerkrieg** [comment]
> I have read almost all changes and this PR is generally Ok.

Except the fact the GCC does not compile, for now. I'm waiting for the build to fail to see the preprocessor output

**Comment by kreuzerkrieg** [comment]
> BTW, I found these changes very useful. Especially: forcing functions to have internal linkage, forcing override...

`override` stuff is far from being completed, it has tons of warnings emitted, I temporary disabled the GCC switch, otherwise it was unbearable. Maybe it is a good idea in the end to merge this specific PR without this warning enabled and then start with another one with `override` warning enabled, since it will be time consuming

File: dbms/src/IO/IReadableWriteBuffer.h
Comment (kreuzerkrieg) [reply dbms/src/IO/IReadableWriteBuffer.h]: Will post here the exact warning

**Comment by alexey-milovidov** [comment]
Good idea.

PS. The best way is to add new warnings one by one.

File: dbms/src/IO/IReadableWriteBuffer.h
Comment (kreuzerkrieg) [reply dbms/src/IO/IReadableWriteBuffer.h]: @alexey-milovidov 
Here you go
In file included from /home/user/Development/ClickHouse/dbms/src/IO/MemoryReadWriteBuffer.h:5,
                 from /home/user/Development/ClickHouse/dbms/src/IO/tests/gtest_cascade_and_memory_write_buffer.cpp:9:
/home/user/Development/ClickHouse/dbms/src/IO/IReadableWriteBuffer.h: In static member function ‘static _Res std::_Function_handler<_Res(_ArgTypes ...), _Functor>::_M_invoke(const std::_Any_data&, _ArgTypes&& ...) [with _Res = std::shared_ptr<DB::WriteBuffer>; _Functor = checkHTTPHandlerCase(size_t, size_t)::<lambda(const WriteBufferPtr&)>; _ArgTypes = {const std::shared_ptr<DB::WriteBuffer>&}]’:
/home/user/Development/ClickHouse/dbms/src/IO/IReadableWriteBuffer.h:14:14: error: potential null pointer dereference [-Werror=null-dereference]
   14 |         if (!can_reread)
      |              ^~~~~~~~~~
/home/user/Development/ClickHouse/dbms/src/IO/IReadableWriteBuffer.h:14:14: error: potential null pointer dereference [-Werror=null-dereference]
At global scope:

**Comment by stavrolia** [comment]
Discussed in TG that @kreuzerkrieg would add warnings outside this PR.

**Pull request closed without merging**
